### PR TITLE
Custom form elements that don't suck

### PR DESCRIPTION
**Pull request by r3dDoX**

### CODE REVIEW

File: drafts/element-internals-api.md [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+Updating the validity does not make a lot of sense yet, but later if we would support things
+like `required` we could take the native checkbox validation messages. This means when trying to
+submit the form our component would report the validity of the native checkbox.
+
+## Focus events
+
+Many 3rd party form libraries allow you to run validations on `change` or on `blur`. This means our
+web component needs to dispatch those events. For our `switch` component, we already dispatch
+a `change` event, and we won't have to worry about focus/blur events. Since we have a native input
+in our ShadowDOM our component will dispathc focus and blur events when the native input receives or
```
Comment (felixmokross) [inline drafts/element-internals-api.md:159]: I think there should be a comma after ShadowDOM

File: drafts/element-internals-api.md [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+* Start using ElementInternals in your web component
+* Ensure value and validity
+* Focus events
+
+## Start using ElementInternals in your web component
+
+There is essentially two things we have to do, to use the ElementInternals API:
+
+* Flag the web component as being `formAssociated`
+* Inject the ElementInterals in the constructor of our web component
```
Comment (felixmokross) [inline drafts/element-internals-api.md:47]: I think this step could be described a bit more in detail. With "Inject" I associated constructor injection, i.e. that it gets passed as a parameter to the constructor. However below in the code sample I learned that we actually call the .attachInternals function to get the internals.

File: drafts/element-internals-api.md [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+domain: campzulu.hashnode.dev
+tags: web-development, lit, web-components, shadow-dom, form
+cover: https://cdn.hashnode.com/res/hashnode/image/stock/unsplash/jJT2r2n7lYA/upload/76adc03462239d3b3e6cb53f44b60b20.jpeg?w=1600&h=840&fit=crop&crop=entropy&auto=compress,format&format=webp
+publishAs: r3dDoX
+hideFromHashnodeCommunity: false
+ignorePost: true
+---
+
+In the world of web development, creating custom form elements has always been a bit of a challenge.
+However, with the introduction of web components and the use of the ElementInternals API, this task
```
Comment (felixmokross) [inline drafts/element-internals-api.md:12]: What I found confusing about ElementInternals, is that they are specifically made for form elements, although the name does not say anything about forms. Maybe we could add a note to that (when first reading the post, I thought ElementInternals could be used for other things too)

**Review by felixmokross — APPROVED** [review]

File: drafts/element-internals-api.md (hunk 3 above)
Comment (r3dDoX) [reply drafts/element-internals-api.md:12]: I called it "attach" now. This way we use the same naming. There is not much more details to go into without going into Browser internals. So I hope this makes it clear enough

File: drafts/element-internals-api.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (r3dDoX) [reply drafts/element-internals-api.md:159]: fixed

Comment (r3dDoX) [reply drafts/element-internals-api.md:159]: Sounds right

File: drafts/element-internals-api.md (hunk 2 above)
Comment (r3dDoX) [reply drafts/element-internals-api.md:47]: I called it "attach" now. This way we use the same naming. There is not much more details to go into without going into Browser internals. So I hope this makes it clear enough

**Pull request merged**